### PR TITLE
Adding Filtered Index support to Python bindings

### PR DESCRIPTION
**Pull request by daxpryce**
- [x] Does this PR have a descriptive title that could go in our release notes?
- [ ] Does this PR add any new dependencies?
- [x] Does this PR modify any existing APIs?
   - [x] Is the change to the API backwards compatible?
- [x] Should this result in any changes to our documentation, either updating existing docs or adding new ones?
 
#### What does this implement/fix? Briefly explain your changes.
Allows for the building of and searching of a filtered memory index, by providing a simple list[list[Hashable]] mapping at build time, and enabling it with a flag on load.

#### Any other comments?
I am not totally confident I know the full set of expectations, behaviors, and bounding edge cases for test in this particular scenario. @harsha-simhadri if you could explain precisely what edge cases I should properly test for other than the basics in this PR, I would be deeply appreciative!

### CODE REVIEW

File: python/src/_builder.py
Comment (daxpryce) [inline python/src/_builder.py:178]: I'm not even confident this is the right signature, since it's just being serialized to a file anyway. Should I change this to `Optional[list[list[str]]]`?

File: python/src/_common.py
Comment (daxpryce) [inline python/src/_common.py:214]: this fixes a bug whose number escapes me

**Comment by daxpryce** [comment]
Closes #455

File: python/src/_common.py
Comment (daxpryce) [reply python/src/_common.py:214]: #455

File: python/src/_builder.py
Comment (bryantower) [reply python/src/_builder.py:178]: It should probably just be str, why do you think it should be Hashable?

**Review by bryantower — COMMENTED** [review]
It all looks pretty good to me. Did you want to wait on a comment from Harsha? I can approve this once we resolve the Hashable vs str issue.

**Review by bryantower — APPROVED** [review]

**Pull request merged**
